src/se/impl/scheduler_impl.hpp

```diff
@@ -0,0 +1,193 @@
+      checkLocked();
+      return std::upper_bound(
+          tasks_.begin(), tasks_.end(), tp, [](const auto &l, const auto &r) {
+            return l < (r.created + r.timeout);
+          });
+    }
+
+    void insert(TaskContainer::const_iterator after, TimedTask &&t) {
+      checkLocked();
+      tasks_.insert(after, std::move(t));
+    }
+
+    bool extractExpired(TimedTask &task) {
+      std::lock_guard lock(tasks_cs_);
+      const Timepoint before = now();
+      if (!tasks_.empty()) {
+        auto &first_task = tasks_.front();
+        const auto timepoint = first_task.created + first_task.timeout;
+        if (timepoint <= before) {
+          task = std::move(first_task);
+          tasks_.pop_front();
+          is_busy_ = true;
+          return true;
+        }
+      }
+      is_busy_ = false;
+      return false;
+    }
+
+    ///@returns time duration from now till first task will be executed
+    std::chrono::microseconds untilFirst() const {
+      std::lock_guard lock(tasks_cs_);
+      const auto before = now();
+      if (!tasks_.empty()) {
+        const auto &first = tasks_.front();
+        const auto timepoint = first.created + first.timeout;
+        if (timepoint > before) {
+          return std::chrono::duration_cast<std::chrono::microseconds>(
+              timepoint - before);
+        }
+
+        return std::chrono::microseconds(0ull);
+      }
+      return std::chrono::minutes(10ull);
+    }
+
+    void add(TimedTask &&task) {
+      assert(!tasks_cs_.try_lock());
+      if (task.timeout == std::chrono::microseconds(0ull)) {
+        is_busy_ = true;
+      }
+
+      insert(after(task.created + task.timeout), std::move(task));
+      event_.set();
+    }
+
+   public:
+    // Disable copying
+    SchedulerBase(const SchedulerBase &) = delete;
+    SchedulerBase &operator=(const SchedulerBase &) = delete;
+
+    // Disable moving
+    SchedulerBase(SchedulerBase &&) = delete;
+    SchedulerBase &operator=(SchedulerBase &&) = delete;
+
+    SchedulerBase() : is_busy_(false) {
+      proceed_.test_and_set();
+    }
+
+    uint32_t process() {
+      id_ = std::this_thread::get_id();
+      TimedTask task{};
+      do {
+        if (extractExpired(task)) {
+          try {
+            if (task.task) {
+              if (!task.predic) {
+                task.task();
+              } else if (task.predic()) {
+                task.task();
+                std::lock_guard lock(tasks_cs_);
+                task.created = now();
+                add(std::move(task));
+              }
+            }
+          } catch (...) {
+          }
+        } else {
+          event_.wait(untilFirst());
+        }
+
+      } while (proceed_.test_and_set());
+      return 0;
+    }
+
+    void dispose(bool wait_for_release = true) override {
+      proceed_.clear();
+      event_.set();
+    }
+
+    bool isBusy() const override {
+      std::lock_guard lock(tasks_cs_);
+      return is_busy_;
```

Comment:
```suggestion
      std::lock_guard lock(tasks_cs_);
      return is_busy_;
```
Why is accessing is_busy_ require locking tasks_cs_? I suspect this relation can be easily broken with future refactors since it's pretty unintuitive, what about SafeObject?